### DISCUSSION TITLE
Build Louvre and SRM as subprojects

### DISCUSSION THREAD

**Discussion by ErikReider** [CuarzoSoftware/Louvre#23] [General]
It would be nice to compile and use Louvre and SRM as a subproject without needing to install them (for debugging purposes). Is this something that's on your radar? :)

**Comment by ehopperdietzel** [comment]
With SRM, you can do that, but with Louvre, this is more challenging because the backends and assets are installed in different directories and loaded at runtime. When you only compile, the meson build directory does not have the structure Louvre needs, so as you may have noticed, it doesn't work.

I believe the best option would be to continue installing but with a different prefix:

```bash
$ meson setup build --prefix=/path/to/install
```

I am also considering adding an option for linking the backends statically, but this would be available later on.

**Comment by EBADBEEF** [comment]
I am also interested in this. It helps for two reasons:
- iteration time in development is improved
- new developers/users casually trying out the project are much more likely to be successful

I was able to build and install Louvre in a local prefix using SRM as a submodule with these two changes:
- https://github.com/CuarzoSoftware/SRM/pull/13
- https://github.com/CuarzoSoftware/Louvre/pull/58

For searching for backends, I think we could use build_rpath and install_rpath meson options to set the default path for looking up backends. Then you could dlopen a relative path.

It would also be good if it only required one environment variable to override the search path (instead of both LOUVRE_BACKENDS_PATH and LOUVRE_GRAPHIC_BACKEND/LOUVRE_INPUT_BACKEND). The issue is due to mismatch between build folder layout and install folder layout.

**Reply by ehopperdietzel** [reply]
Thank you again! I've just merged both PRs :)

> It would also be good if it only required one environment variable to override the search path (instead of both LOUVRE_BACKENDS_PATH and LOUVRE_GRAPHIC_BACKEND/LOUVRE_INPUT_BACKEND).

`BACKENDS_INSTALL_PATH = join_paths(get_option('prefix'), get_option('libdir'), 'Louvre/backends')`

This line in `meson.build` defines the backends path, which can be modified using the `prefix` and `libdir` options. However, I think these need to be absolute, I haven't tried relative paths. This path is also hardcoded as the default when the library is built.

In Louvre 2.0.0, only the **LOUVRE_BACKENDS_PATH** env is required to modify the backends path. Louvre will attempt to use the Wayland backend if **WAYLAND_DISPLAY** is set, and fall back to DRM otherwise.

**LOUVRE_GRAPHIC_BACKEND** and **LOUVRE_INPUT_BACKEND** are only meant to specify the backend name, such as `drm` or `wayland`, to correctly select the `backend.so` file. The **Backends Configuration** section [here](https://cuarzosoftware.github.io/Louvre/md_md__environment.html) explains it in detail.

Lastly, it is also possible to load the backends programmatically before calling `compositor.start()`, using [LCompositor::loadGraphicBackend()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_compositor.html#a00c61d589cffdbd44ff06a937b188941) and [LCompositor::loadInputBackend()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_compositor.html#a36e5e82a78cc618368d4fdf0bb2d6b4d).
